application/src/main/java/org/togetherjava/tjbot/commands/moderation/scam/ScamBlocker.java

```diff
@@ -0,0 +1,364 @@
+    }
+
+    private void quarantineAuthor(@NotNull Guild guild, @NotNull Member author,
+            @NotNull SelfUser bot) {
+        String reason = "User posted scam that was automatically detected";
+
+        actionsStore.addAction(guild.getIdLong(), bot.getIdLong(), author.getIdLong(),
+                ModerationAction.QUARANTINE, null, reason);
+
+        guild
+            .addRoleToMember(author,
+                    ModerationUtils.getQuarantinedRole(guild, config).orElseThrow())
+            .reason(reason)
+            .queue();
+    }
+
+    private void reportScamMessage(@NotNull GuildMessageReceivedEvent event,
+            @NotNull String reportTitle, @NotNull Collection<? extends ActionRow> actionRows) {
+        Guild guild = event.getGuild();
+        Optional<TextChannel> reportChannel = getReportChannel(guild);
+        if (reportChannel.isEmpty()) {
+            logger.warn(
+                    "Unable to report a scam message, did not find a report channel matching the configured pattern '{}' for guild '{}'",
+                    reportChannelPattern, guild.getName());
+            return;
+        }
+
+        User author = event.getAuthor();
+        MessageEmbed embed =
+                new EmbedBuilder().setDescription(event.getMessage().getContentStripped())
+                    .setTitle(reportTitle)
+                    .setAuthor(author.getAsTag(), null, author.getAvatarUrl())
+                    .setTimestamp(event.getMessage().getTimeCreated())
+                    .setColor(AMBIENT_COLOR)
+                    .setFooter(author.getId())
+                    .build();
+        Message message = new MessageBuilder().setEmbeds(embed).setActionRows(actionRows).build();
+
+        reportChannel.orElseThrow().sendMessage(message).queue();
+    }
+
+    private void dmUser(@NotNull GuildMessageReceivedEvent event) {
+        dmUser(event.getGuild(), event.getAuthor().getIdLong(), event.getJDA());
+    }
+
+    private void dmUser(@NotNull Guild guild, long userId, @NotNull JDA jda) {
+        String dmMessage =
+                """
+                        Hey there, we detected that you did send scam in the server %s and therefore put you under quarantine.
+                        This means you can no longer interact with anyone in the server until you have been unquarantined again.
+
+                        If you think this was a mistake (for example, your account was hacked, but you got back control over it),
+                        please contact a moderator or admin of the server.
+                        """
+                    .formatted(guild.getName());
+
+        jda.openPrivateChannelById(userId)
+            .flatMap(channel -> channel.sendMessage(dmMessage))
+            .queue();
+    }
+
+    private Optional<TextChannel> getReportChannel(@NotNull Guild guild) {
+        // Check cache first, then get full list
```

Comment:
pretty sure i copied that from somewhere. so that other location is still faulty then as well.

that said, afaik retrieve checks the cache beforehand first anyways, no? iirc i read that in the doc of other commands, such as retrieveUser.

Reply:
It does, but the point I'm making is that it's getTextChannels not retrieveTextChannels.

Reply:
> It does, but the point I'm making is that it's getTextChannels not retrieveTextChannels.

I understand. My point is, I can just do `retrieveFoo()` and only that, since it will internally first check the cache anyways, no?

Reply:
Correc

Reply:
Checked the code and we both overlooked something in this discussion. The code is correct and cant be improved imo:

![screen](https://i.imgur.com/Kfttk34.png)

Basically, we dont have the channels ID or name, we have a predicate matching the name. So we need to check the list of all channels against the predicate for any-match. but `getTextChannels()` (no cache) will always create a full new list, it says itself that one should first check the cache:

![screen2](https://i.imgur.com/Pgsyk8n.png)

Reply:
I didn't overlook anything, you are the one overlooking, still after me explaining to you a few times.

`getTextChannels = new List(getTextChannelCache())` (not exactly, but 90% the same)

getTextChannels **doesn't retrieve**, it would be called *retrieveTextChannels* then.

use getTextChannelCache

Reply:
I see, thats fair. Now it clicked.

Reply:
And I can't find were you copied it from, don't see any class call getTextChannels even. So maybe misremembering or copied it from an outdated PR? Not sure

Reply:
Cant find it either anymore. Anyways, fixed.